src/Conversion/ONNXToKrnl/Math/LRN.cpp

```diff
@@ -0,0 +1,143 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+struct ONNXLRNOpLowering : public ConversionPattern {
+  ONNXLRNOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXLRNOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXLRNOpAdaptor operandAdaptor(operands);
+    ONNXLRNOp lrnOp = llvm::cast<ONNXLRNOp>(op);
+    auto loc = op->getLoc();
+
+    ONNXLRNOpShapeHelper shapeHelper(&lrnOp, &rewriter);
+
+    auto shapecomputed = shapeHelper.Compute(operandAdaptor);
+    (void)shapecomputed;
+    assert(!failed(shapecomputed) && "expected to succeed");
+
+    auto resultOperand = lrnOp.Y();
+    auto outputMemRefType = convertToMemRefType(*op->result_type_begin());
+    auto outputMemRefShape = outputMemRefType.getShape();
+    auto elementType = outputMemRefType.getElementType();
+    int64_t outputRank = outputMemRefShape.size();
+
+    Value input = operandAdaptor.X();
+    float biasLit = lrnOp.bias().convertToFloat();
+    float alphaLit = lrnOp.alpha().convertToFloat();
+    float betaLit = lrnOp.beta().convertToFloat();
+    float sizeLit = (float)lrnOp.size();
+    auto f32Type = FloatType::getF32(rewriter.getContext());
+    Value biasValue = emitConstantOp(rewriter, loc, f32Type, biasLit);
+    Value alphaDivSizeValue =
+        emitConstantOp(rewriter, loc, f32Type, alphaLit / sizeLit);
+    Value betaValue = emitConstantOp(rewriter, loc, f32Type, betaLit);
+
+    Value alloc = insertAllocAndDeallocSimple(
+        rewriter, op, outputMemRefType, loc, shapeHelper.dimsForOutput(0));
+
+    BuildKrnlLoop outputLoops(rewriter, loc, outputRank);
+    outputLoops.createDefineOp();
+    outputLoops.pushAllBounds(shapeHelper.dimsForOutput(0));
+    outputLoops.createIterateOp();
+    rewriter.setInsertionPointToStart(outputLoops.getIterateBlock());
+
+    // Insert computation of square_sum.
+    // square_sum[n, c, d1, ..., dk] = sum(X[n, i, d1, ..., dk] ^ 2),
+    // where max(0, c - floor((size - 1) / 2)) <= i
+    // and i<= min(C - 1, c + ceil((size - 1) / 2)).
+
+    // Get a child IndexExpr context.
+    IndexExprContext childContext(shapeHelper.context);
+
+    // Compute the lower bound and upper bound for square_sum.
+    const int loopIndexForC = 1;
+    Value cValue = outputLoops.getInductionVar(loopIndexForC);
+    IndexExpr cIE = childContext.createLoopInductionIndex(cValue);
+    IndexExpr sizeIE =
+        childContext.createDimIndexFromShapedType(input, loopIndexForC);
+
+    SmallVector<IndexExpr, 2> lbMaxList;
+    lbMaxList.emplace_back(childContext.createLiteralIndex(0));
+    lbMaxList.emplace_back(
+        cIE - (sizeIE - 1).floorDiv(childContext.createLiteralIndex(2)));
```

Comment:
Minor: could I ask you to add the IndexExpr::floorDiv(int64_t b) and same for ceilDiv? Looks like I omitted to add it. Looks like now would be a good time to add it.

```
IndexExpr IndexExpr::operator*(int64_t const b) const {
  IndexExpr bIndex = getContext().createLiteralIndex(b);
  return *this * bIndex;
}
```

Reply:
yes, I can add them.  As I mentioned in the later comment: in another patch.